### PR TITLE
Add new public APIs for safely manipulating managed pointers

### PR DESCRIPTION
**Pull request by kg**
I still need to adjust this PR so that it won't break blazor when we land it, but I've been experimenting with an updated version of blazor that works great on top of it.

* Update WasmRootBuffer so it can point into implicitly-rooted stack memory
* Add mono_wasm_with_hazard_buffer API you can use to request a root buffer that automatically pins (not just roots) all of the pointers inside of it, allowing you to manipulate them freely
* Add mono_wasm_with_pinned_object API that temporarily pins the contents of a root and then passes you the raw pointer
* Expose stackSave/Restore/Alloc emscripten APIs as part of Module
* Fix GetCallSignatureRef not properly marshaling its return value
* Add new API that converts an arbitrary managed array into a JS TypedArray view of a given type (or copy of the data)
* Introduce new safe variant of blazor's invoke API
* Add experimental APIs for safely manipulating fields of managed objects

### CODE REVIEW

**Comment by pavelsavara** [comment]
What are the use-cases at this point ?

File: src/mono/wasm/wasm.proj [hunk 1]
```diff
@@ -130,7 +129,10 @@
       <_EmccLinkFlags Include="-s ALLOW_MEMORY_GROWTH=1" />
       <_EmccLinkFlags Include="-s NO_EXIT_RUNTIME=1" />
       <_EmccLinkFlags Include="-s FORCE_FILESYSTEM=1" />
-      <_EmccLinkFlags Include="-s EXPORTED_RUNTIME_METHODS=&quot;['FS','print','ccall','cwrap','setValue','getValue','UTF8ToString','UTF8ArrayToString','FS_createPath','FS_createDataFile','removeRunDependency','addRunDependency', 'FS_readFile']&quot;"  />
+      <_EmccLinkFlags Include="-s EXPORTED_RUNTIME_METHODS=&quot;['FS','print','ccall','cwrap','setValue','getValue','UTF8ToString','UTF8ArrayToString','FS_createPath','FS_createDataFile','removeRunDependency','addRunDependency','stackSave','stackRestore','stackAlloc','FS_readFile']&quot;"  />
+      <_EmccCommonFlags Condition="'$(MonoWasmThreads)' == 'true'" Include="-s USE_PTHREADS=1" />
```
Comment (pavelsavara) [inline src/mono/wasm/wasm.proj:133]: thread relates stuff is unrelated? I think it should be in another PR

**Comment by kg** [comment]
> What are the use-cases at this point ?

Same as before, current blazor is not GC safe. This allows patching the few remaining unmarshalled calls to be GC safe

**Comment by lambdageek** [comment]
@kg something went wrong with the merge here. looks like you're reverting a lot of unrelated work

**Comment by kg** [comment]
> @kg something went wrong with the merge here. looks like you're reverting a lot of unrelated work

Thanks, I'll rebase it again. It took me a while to get it to rebase at all.

Any idea why it keeps automatically re-requesting review?

**Comment by maraf** [comment]
> Any idea why it keeps automatically re-requesting review?

Probably because of the merge damage and change in a lot of areas?
